File: app/src/main/java/dev/ragnarok/fenrir/model/Dialog.java
package dev.ragnarok.fenrir.model;

import static dev.ragnarok.fenrir.util.Utils.firstNonEmptyString;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.api.model.Identificable;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Objects;
import dev.ragnarok.fenrir.util.Utils;

public class Dialog implements Identificable, Parcelable {

    public static final Creator<Dialog> CREATOR = new Creator<Dialog>() {
        @Override
        public Dialog createFromParcel(Parcel in) {
            return new Dialog(in);
        }

        @Override
        public Dialog[] newArray(int size) {
            return new Dialog[size];
        }
    };
    private int peerId;
    private String title;
    private int unreadCount;
    private String photo50;
    private String photo100;
    private String photo200;
    private Message message;
    private Owner interlocutor;
    private int lastMessageId;
    private int inRead;
    private int outRead;
    private boolean isGroupChannel;
    private int major_id;
    private int minor_id;

    public Dialog() {

    }

    protected Dialog(Parcel in) {
        peerId = in.readInt();
        title = in.readString();
        unreadCount = in.readInt();
        photo50 = in.readString();
        photo100 = in.readString();
        photo200 = in.readString();
        message = in.readParcelable(Message.class.getClassLoader());

        boolean interlocutorIsNull = in.readInt() == 1;
        if (!interlocutorIsNull) {
            int ownerType = in.readInt();
            interlocutor = in.readParcelable(ownerType == OwnerType.COMMUNITY
                    ? Community.class.getClassLoader() : User.class.getClassLoader());
        }

        lastMessageId = in.readInt();
        inRead = in.readInt();
        outRead = in.readInt();
        major_id = in.readInt();
        minor_id = in.readInt();
        isGroupChannel = in.readInt() == 1;
    }

    public int getPeerId() {
        return peerId;
    }

    public Dialog setPeerId(int peerId) {
        this.peerId = peerId;
        return this;
    }

    public String getTitle() {
        if (Peer.isUser(peerId)) {
            String custom = Settings.get().other().getUserNameChanges(peerId);
            if (!Utils.isEmpty(custom)) {
                return custom;
            }
        }
        return title;
    }

    public Dialog setTitle(String title) {
        this.title = title;
        return this;
    }

    public int getUnreadCount() {
        return unreadCount;
    }

    public Dialog setUnreadCount(int unreadCount) {
        this.unreadCount = unreadCount;
        return this;
    }

    public String getPhoto50() {
        return photo50;
    }

    public Dialog setPhoto50(String photo50) {
        this.photo50 = photo50;
        return this;
    }

    public String getPhoto100() {
        return photo100;
    }

    public Dialog setPhoto100(String photo100) {
        this.photo100 = photo100;
        return this;
    }

    public String getPhoto200() {
        return photo200;
    }

    public Dialog setPhoto200(String photo200) {
        this.photo200 = photo200;
        return this;
    }

    public boolean isChat() {
        return Peer.isGroupChat(peerId);
    }

    public boolean isUser() {
        return Peer.isUser(peerId);
    }

    public boolean isGroup() {
        return Peer.isGroup(peerId);
    }

    public Message getMessage() {
        return message;
    }

    public Dialog setMessage(Message message) {
        this.message = message;
        return this;
    }

    public Owner getInterlocutor() {
        return interlocutor;
    }

    public Dialog setInterlocutor(Owner interlocutor) {
        this.interlocutor = interlocutor;
        return this;
    }

    public int getLastMessageId() {
        return lastMessageId;
    }

    public Dialog setLastMessageId(int lastMessageId) {
        this.lastMessageId = lastMessageId;
        return this;
    }

    public boolean isLastMessageOut() {
        return message != null && message.isOut();
    }

    @ChatAction
    public Integer getLastMessageAction() {
        return message == null ? null : message.getAction();
    }

    public boolean hasForwardMessages() {
        return getForwardMessagesCount() > 0;
    }

    public int getForwardMessagesCount() {
        return message == null ? 0 : message.getForwardMessagesCount();
    }

    public boolean hasAttachments() {
        return message != null && message.isHasAttachments();
    }

    public String getLastMessageBody() {
        return Objects.isNull(message) ? "..." : message.getCryptStatus() == CryptStatus.DECRYPTED ? message.getDecryptedBody() : message.getBody();
    }

    @NonNull
    public String getSenderShortName(@NonNull Context context) {
        String targerText = null;
        if (interlocutor instanceof User) {
            targerText = ((User) interlocutor).getFirstName();
        } else if (interlocutor instanceof Community) {
            targerText = ((Community) interlocutor).getName();
        }

        return targerText == null ? context.getString(R.string.unknown_first_name) : targerText;
    }

    public String getDisplayTitle(@NonNull Context context) {
        switch (Peer.getType(peerId)) {
            case Peer.USER:
            case Peer.GROUP:
                return interlocutor == null ? context.getString(R.string.unknown_first_name) + " " + context.getString(R.string.unknown_last_name) : interlocutor.getFullName();
            case Peer.CHAT:
                return title;
            default:
                throw new IllegalStateException("Unknown peer id: " + peerId);
        }
    }

    public long getLastMessageDate() {
        return message == null ? 0 : message.getDate();
    }

    public String getImageUrl() {
        if (Peer.getType(peerId) == Peer.CHAT) {

            //if (isEmpty(img) && interlocutor != null) {
            //img = interlocutor.getMaxSquareAvatar();
            // }

            return firstNonEmptyString(photo200, photo100, photo50);
        }
        return interlocutor == null ? null : interlocutor.getMaxSquareAvatar();
    }

    public int getInRead() {
        return inRead;
    }

    public Dialog setInRead(int inRead) {
        this.inRead = inRead;
        return this;
    }

    public boolean isLastMessageRead() {
        return isLastMessageOut() ? getLastMessageId() <= outRead : getLastMessageId() <= inRead;
    }

    public int getOutRead() {
        return outRead;
    }

    public Dialog setOutRead(int outRead) {
        this.outRead = outRead;
        return this;
    }

    public boolean isGroupChannel() {
        return isGroupChannel;
    }

    public Dialog setGroupChannel(boolean groupChannel) {
        isGroupChannel = groupChannel;
        return this;
    }

    public int getMajor_id() {
        return major_id;
    }

    public Dialog setMajor_id(int major_id) {
        this.major_id = major_id;
        return this;
    }

    public int getMinor_id() {
        if (minor_id == 0) {
            return getLastMessageId();
        }
        return minor_id;
    }

    public Dialog setMinor_id(int minor_id) {
        this.minor_id = minor_id;
        return this;
    }

    @Override
    public int getId() {
        return peerId;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(peerId);
        dest.writeString(title);
        dest.writeInt(unreadCount);
        dest.writeString(photo50);
        dest.writeString(photo100);
        dest.writeString(photo200);
        dest.writeParcelable(message, flags);

        dest.writeInt(interlocutor == null ? 1 : 0);
        if (interlocutor != null) {
            dest.writeInt(interlocutor.getOwnerType());
            dest.writeParcelable(interlocutor, flags);
        }

        dest.writeInt(lastMessageId);
        dest.writeInt(inRead);
        dest.writeInt(outRead);
        dest.writeInt(major_id);
        dest.writeInt(minor_id);
        dest.writeInt(isGroupChannel ? 1 : 0);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Dialog.kt
package dev.ragnarok.fenrir.model

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.*
import dev.ragnarok.fenrir.api.model.Identificable
import dev.ragnarok.fenrir.settings.Settings.get
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString

class Dialog : Identificable, Parcelable {
    var peerId = 0
        private set
    private var title: String? = null
    var unreadCount = 0
        private set
    var photo50: String? = null
        private set
    var photo100: String? = null
        private set
    var photo200: String? = null
        private set
    var message: Message? = null
        private set
    var interlocutor: Owner? = null
        private set
    var lastMessageId = 0
        private set
    var inRead = 0
        private set
    var outRead = 0
        private set
    var isGroupChannel = false
        private set
    var major_id = 0
        private set
    private var minor_id = 0

    constructor()
    internal constructor(`in`: Parcel) {
        peerId = `in`.readInt()
        title = `in`.readString()
        unreadCount = `in`.readInt()
        photo50 = `in`.readString()
        photo100 = `in`.readString()
        photo200 = `in`.readString()
        message = `in`.readTypedObjectCompat(Message.CREATOR)
        val interlocutorIsNull = `in`.getBoolean()
        if (!interlocutorIsNull) {
            interlocutor = Owner.readOwnerFromParcel(`in`)
        }
        lastMessageId = `in`.readInt()
        inRead = `in`.readInt()
        outRead = `in`.readInt()
        major_id = `in`.readInt()
        minor_id = `in`.readInt()
        isGroupChannel = `in`.getBoolean()
    }

    fun setPeerId(peerId: Int): Dialog {
        this.peerId = peerId
        return this
    }

    fun getTitle(): String? {
        return title
    }

    fun setTitle(title: String?): Dialog {
        this.title = title
        return this
    }

    fun setUnreadCount(unreadCount: Int): Dialog {
        this.unreadCount = unreadCount
        return this
    }

    fun setPhoto50(photo50: String?): Dialog {
        this.photo50 = photo50
        return this
    }

    fun setPhoto100(photo100: String?): Dialog {
        this.photo100 = photo100
        return this
    }

    fun setPhoto200(photo200: String?): Dialog {
        this.photo200 = photo200
        return this
    }

    val isChat: Boolean
        get() = Peer.isGroupChat(peerId)
    val isUser: Boolean
        get() = Peer.isUser(peerId)
    val isGroup: Boolean
        get() = Peer.isGroup(peerId)

    val isContact: Boolean
        get() = Peer.isContactChat(peerId)

    fun setMessage(message: Message?): Dialog {
        this.message = message
        return this
    }

    fun setInterlocutor(interlocutor: Owner?): Dialog {
        this.interlocutor = interlocutor
        return this
    }

    fun setLastMessageId(lastMessageId: Int): Dialog {
        this.lastMessageId = lastMessageId
        return this
    }

    val isLastMessageOut: Boolean
        get() = message?.isOut == true

    @ChatAction
    val lastMessageAction: Int
        get() = message?.action ?: ChatAction.NO_ACTION

    fun hasForwardMessages(): Boolean {
        return forwardMessagesCount > 0
    }

    val forwardMessagesCount: Int
        get() = message?.forwardMessagesCount.orZero()

    fun hasAttachments(): Boolean {
        return message?.isHasAttachments == true
    }

    val lastMessageBody: String?
        get() = if (message == null) "..." else if (message?.cryptStatus == CryptStatus.DECRYPTED) message?.decryptedBody else message?.body

    fun getSenderShortName(context: Context): String {
        var targerText: String? = null
        if (interlocutor is User) {
            targerText = (interlocutor as User).firstName
        } else if (interlocutor is Community) {
            targerText = (interlocutor as Community).fullName
        }
        return targerText ?: context.getString(R.string.unknown_first_name)
    }

    fun getDisplayTitle(context: Context): String? {
        return when (Peer.getType(peerId)) {
            Peer.USER, Peer.GROUP -> {
                val custom = get().other().getUserNameChanges(peerId)
                if (custom.nonNullNoEmpty()) {
                    return custom
                }
                if (interlocutor == null && title.isNullOrEmpty()) context.getString(R.string.unknown_first_name) + " " + context.getString(
                    R.string.unknown_last_name
                )
                else if (title.nonNullNoEmpty()) {
                    title
                } else interlocutor?.fullName
            }
            Peer.CHAT, Peer.CONTACT -> title
            else -> throw IllegalStateException("Unknown peer id: $peerId")
        }
    }

    fun getDisplayTitle(): String? {
        return when (Peer.getType(peerId)) {
            Peer.USER, Peer.GROUP -> {
                val custom = get().other().getUserNameChanges(peerId)
                if (custom.nonNullNoEmpty()) {
                    return custom
                }
                if (interlocutor == null && title.isNullOrEmpty()) null
                else if (title.nonNullNoEmpty()) {
                    title
                } else interlocutor?.fullName
            }
            Peer.CHAT, Peer.CONTACT -> title
            else -> throw IllegalStateException("Unknown peer id: $peerId")
        }
    }

    val lastMessageDate: Long
        get() = message?.date.orZero()

    //if (isEmpty(img) && interlocutor != null) {
    //img = interlocutor.getMaxSquareAvatar();
    // }
    val imageUrl: String?
        get() {
            if (Peer.getType(peerId) == Peer.CHAT || Peer.getType(peerId) == Peer.CONTACT) {

                //if (isEmpty(img) && interlocutor != null) {
                //img = interlocutor.getMaxSquareAvatar();
                // }
                return firstNonEmptyString(photo200, photo100, photo50)
            }
            return interlocutor?.maxSquareAvatar
        }

    fun setInRead(inRead: Int): Dialog {
        this.inRead = inRead
        return this
    }

    val isLastMessageRead: Boolean
        get() = if (isLastMessageOut) lastMessageId <= outRead else lastMessageId <= inRead

    fun setOutRead(outRead: Int): Dialog {
        this.outRead = outRead
        return this
    }

    fun setGroupChannel(groupChannel: Boolean): Dialog {
        isGroupChannel = groupChannel
        return this
    }

    fun setMajor_id(major_id: Int): Dialog {
        this.major_id = major_id
        return this
    }

    fun getMinor_id(): Int {
        return if (minor_id == 0) {
            lastMessageId
        } else minor_id
    }

    fun setMinor_id(minor_id: Int): Dialog {
        this.minor_id = minor_id
        return this
    }

    override fun getObjectId(): Int {
        return peerId
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(peerId)
        dest.writeString(title)
        dest.writeInt(unreadCount)
        dest.writeString(photo50)
        dest.writeString(photo100)
        dest.writeString(photo200)
        dest.writeTypedObjectCompat(message, flags)
        dest.writeInt(if (interlocutor == null) 1 else 0)
        interlocutor.requireNonNull {
            Owner.writeOwnerToParcel(it, dest, flags)
        }
        dest.writeInt(lastMessageId)
        dest.writeInt(inRead)
        dest.writeInt(outRead)
        dest.writeInt(major_id)
        dest.writeInt(minor_id)
        dest.putBoolean(isGroupChannel)
    }

    companion object CREATOR : Parcelable.Creator<Dialog> {
        override fun createFromParcel(parcel: Parcel): Dialog {
            return Dialog(parcel)
        }

        override fun newArray(size: Int): Array<Dialog?> {
            return arrayOfNulls(size)
        }
    }
}